---
editor:
    render-on-save: true
bibliography: references.bib
csl: ieee.csl
---

## Experimental setup{#sec-experiment}

The present work is designed to analyze time-series data of relative abundance
of multiple microbial lineages uniquely identified by a DNA barcode
[@kinsler2020; @ascensao2023]. In these competition assays, an ensemble of
genotypes is pooled together with an unlabeled reference strain that, initially,
represents the vast majority ($\geq 90\%$) of the cells in the culture (see
schematic in @fig-01(A)). Furthermore, a fraction of labeled genotypes
equivalent to the unlabeled reference strain---hereafter defined as *neutral*
lineages---are spiked in at a relatively high abundance ($\approx 3-5\%$). The
rest of the culture is left for the ensemble of genotypes of interest.
<span style="color: #912E27;">
This experimental design in which the barcodes of interest represent a small
fraction of the culture serves three main purposes: First, the model used to
infer the relative fitness differences between genotypes is valid in the regime
where the genotype frequency is significantly smaller than one (see
@sec-fitness_model). Second, potential problems with frequency-dependent
selection are minimized as long as the frequency of each genotype remains small.
Third, having a single genotype dominating the culture standardizes the
environment experienced by all genotypes. This is because variations in the
chemistry of the environment are effectively dictated by the dominating
genotype, allowing for reproducible growth-dilution cycles.
</span>

To determine the relative fitness of the ensemble of genotypes, a series of
growth-dilution cycles are performed on either a single or multiple
environments. In other words, the cultures are grown for some time; then, an
aliquot is inoculated into fresh media for the next growth cycle. This process
is repeated for roughly 4-7 cycles, depending on the initial abundances of the
mutants and their relative growth rates. The DNA barcodes are sequenced at the
end of each growth cycle to quantify the relative abundance of each of the
barcodes. We point the reader to @kinsler2020 for specific details on these
assays for *S. cerevisiae* and to @ascensao2023 for equivalent assays for *E.
coli*. @fig-01(B) presents a typical barcode trajectory where the black
trajectories represent the so-called *neutral lineages*, genetically equivalent
to the untagged ancestor strain that initially dominates the culture. These
spiked-in neutral lineages simplify the inference problem since the fitness
metric of all relevant barcodes is quantified with respect to these
barcodes---thus referred to as *relative fitness*.

![**Typical competitive fitness experiment**. (A) Schematic of the typical
experimental design to determine the competitive fitness of an ensemble of
barcoded genotypes. Genotypes are pooled together and grown over multiple
growth-dilution cycles. At the end of each cycle, a sample is processed to
generate a library for amplicon sequencing. (B) Typical barcode trajectory
dataset. From each time point, the relative frequency of each barcode is
determined from the total number of reads. Shades of blue represent different
relative fitness. Darker gray lines define the typical trajectory of neutral
lineages. 
<span style="color: #912E27;">
(C) Sources of uncertainty accounted for by our method. The Bayesian
model fit to the data propagates uncertainties from the categories schematically
depicted to all parameters in the inference.
</span>](./figs/fig01){#fig-01}
  

  
